### PR TITLE
feat: Added IRepository

### PR DESCRIPTION
**Pull request by sammychinedu2ky**
fixes https://github.com/oras-project/oras-dotnet/issues/17

### CODE REVIEW

File: Oras/Interfaces/IDeleter.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using Oras.Models;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Oras.Interfaces
+{
+    /// <summary>
+    /// IDeleter removes content.
+    /// </summary>
+    internal interface IDeleter
```
Comment (Wwwsylvia) [inline Oras/Interfaces/IDeleter.cs:10]: public

File: Oras/Interfaces/IBlobTarget.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Oras.Interfaces
+{
+    /// <summary>
+    /// IBlobTarget is a CAS with the ability to stat and delete its content.
+    /// </summary>
+    internal interface IBlobTarget : ITarget, IResolver, IDeleter, IReferenceFetcher
```
Comment (Wwwsylvia) [inline Oras/Interfaces/IBlobTarget.cs:6]: `BlobStore`?  See https://github.com/oras-project/oras-go/blob/v2.0.2/registry/repository.go#L56-L62. Note that it should not implement `ITarget`.

File: Oras/Interfaces/IManifestTarget.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Oras.Interfaces
+{
+    /// <summary>
+    /// IManifestTarget is a CAS with the ability to stat and delete its content.
+    /// Besides, ManifestTarget provides reference tagging.
+    /// </summary>
+    internal interface IManifestTarget : IBlobTarget, IReferencePusher, ITag
```
Comment (Wwwsylvia) [inline Oras/Interfaces/IManifestTarget.cs:7]: `IManifestStore`?

File: Oras/Interfaces/IRepository.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    /// assume that content.Resolver interface only works for manifests. Tagging a
+    /// blob may be resulted in an `ErrUnsupported` error. However, this interface
+    /// does not restrict tagging blobs.
+    /// Since a repository is an union of the blob and the manifest CASs, all
+    /// operations defined in the `BlobStore` are executed depending on the media
+    /// type of the given descriptor accordingly.
+    /// Furthermore, this interface also provides the ability to enforce the
+    /// separation of the blob and the manifests CASs.
+    /// </summary>
+    internal interface IRepository : ITarget, ITagResolver, IReferenceFetcher, IReferencePusher, IReferrerLister, IDeleter
```
Comment (Wwwsylvia) [inline Oras/Interfaces/IRepository.cs:17]: `ITarget` already implies `ITagResolver`, so `ITagResolver` is redundant here.

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: Oras/Interfaces/ITagLister.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        /// specification.
+        /// References:
+        /// - https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#content-discovery
+        /// - https://docs.docker.com/registry/spec/api/#tags
+        /// See also `Tags()` in this package.
+        /// </summary>
+        /// <param name="last"></param>
+        /// <param name="fn"></param>
+        /// <returns></returns>
+        Task TagsAsync(string last, Func<string[], Task> fn);
```
[the remark this answers is not in the document]
Comment (sammychinedu2ky) [reply Oras/Interfaces/ITagLister.cs:31]: Yh you're right

File: Oras/Interfaces/IReferenceFetcher.cs
[the remark this answers is not in the document]
Comment (sammychinedu2ky) [reply Oras/Interfaces/IReferenceFetcher.cs:1]: I think that would be okay for easier identification and accessibility

File: Oras/Interfaces/ITag.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿using Oras.Models;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Oras.Interfaces
+{
+    /// <summary>
+    /// ITag tags reference tags
+    /// </summary>
+    public interface ITag
```
[the remark this answers is not in the document]
Comment (sammychinedu2ky) [reply Oras/Interfaces/ITag.cs:10]: Anyone is okay, but let me conform to the golang code style

File: Oras/Interfaces/Registry/IReferenceFetcher.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    /// </summary>
+    public interface IReferenceFetcher
+    {
+        /// <summary>
+        /// FetchReferenceAsync fetches the content identified by the reference.
+        /// </summary>
+        /// <param name="reference"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        (Task<Descriptor>, Task<Stream>) FetchReferenceAsync(string reference, CancellationToken cancellationToken = default);
```
Comment (shizhMSFT) [inline Oras/Interfaces/Registry/IReferenceFetcher.cs:19]: Question for you: How do you invoke this method?

Comment (shizhMSFT) [reply Oras/Interfaces/Registry/IReferenceFetcher.cs:19]: You have multiple options:

Option 1:
```csharp
Task<(Descriptor, Stream)> FetchReferenceAsync(string reference, CancellationToken cancellationToken = default);
```

Option 2:
```csharp
Task<Descriptor> FetchReferenceAsync(string reference, out Steam steam CancellationToken cancellationToken = default);
```

Option 3:
```csharp
Task<Stream> FetchReferenceAsync(string reference, out Descriptor descriptor, CancellationToken cancellationToken = default);
```

File: Oras/Interfaces/Registry/ITagLister.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    {
+        /// <summary>
+        /// Tags lists the tags available in the repository.
+        /// Since the returned tag list may be paginated by the underlying
+        /// implementation, a function should be passed in to process the paginated
+        /// tag list.
+        /// `last` argument is the `last` parameter when invoking the tags API.
+        /// If `last` is NOT empty, the entries in the response start after the
+        /// tag specified by `last`. Otherwise, the response starts from the top
+        /// of the Tags list.
```
Comment (Wwwsylvia) [inline Oras/Interfaces/Registry/ITagLister.cs:20]: nit: Should we move the descriptor for the parameters into <param>s?

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: Oras/Interfaces/Registry/IDeleter.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
+﻿using Oras.Models;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Oras.Interfaces.Registry
+{
+    /// <summary>
+    /// IDeleter removes content.
+    /// </summary>
+    public interface IDeleter
```
Comment (Wwwsylvia) [inline Oras/Interfaces/Registry/IDeleter.cs:10]: `IDeleter` should not be under the registry namespace.

File: Oras/Interfaces/Registry/IRepository.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
+﻿namespace Oras.Interfaces.Registry
+{
+    /// <summary>
+    /// Repository is an ORAS target and an union of the blob and the manifest CASs.
+    /// As specified by https://docs.docker.com/registry/spec/api/, it is natural to
+    /// assume that content.Resolver interface only works for manifests. Tagging a
```
Comment (Wwwsylvia) [inline Oras/Interfaces/Registry/IRepository.cs:6]: nit: 
```suggestion
    /// assume that IResolver interface only works for manifests. Tagging a
```

File: Oras/Interfaces/Registry/IRepository.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    /// blob may be resulted in an `ErrUnsupported` error. However, this interface
```
Comment (Wwwsylvia) [inline Oras/Interfaces/Registry/IRepository.cs:7]: There is no `ErrUnsupported` error in the oras-dotnet.

File: Oras/Interfaces/Registry/IRepository.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    /// does not restrict tagging blobs.
+    /// Since a repository is an union of the blob and the manifest CASs, all
+    /// operations defined in the `BlobStore` are executed depending on the media
```
Comment (Wwwsylvia) [inline Oras/Interfaces/Registry/IRepository.cs:10]: IBlobStore? Please review other comments for similar issues.

**Review by Wwwsylvia — APPROVED** [review]
LGTM

**Pull request merged**